numpy/core/_type_aliases.py

```diff
@@ -17,221 +17,87 @@
-                  ('object_', 'object'),
-                  ('cfloat', 'cdouble')]
-    for alias, t in type_pairs:
-        allTypes[alias] = allTypes[t]
-        sctypeDict[alias] = sctypeDict[t]
-    # Remove aliases overriding python types and modules
-    to_remove = ['object', 'int', 'float', 'complex', 'bool', 
-                 'string', 'datetime', 'timedelta', 'bytes', 'str']
-
-    for t in to_remove:
-        try:
-            del allTypes[t]
-            del sctypeDict[t]
-        except KeyError:
-            pass
-
-    # Additional aliases in sctypeDict that should not be exposed as attributes
-    attrs_to_remove = ['ulong', 'long', 'unicode', 'cfloat']
-
-    for t in attrs_to_remove:
-        try:
-            del allTypes[t]
-        except KeyError:
-            pass
-_set_up_aliases()
-
-
-sctypes = {'int': [],
-           'uint': [],
-           'float': [],
-           'complex': [],
-           'others': [bool, object, bytes, str, void]}
-
-
-def _add_array_type(typename, bits):
-    try:
-        t = allTypes['%s%d' % (typename, bits)]
-    except KeyError:
-        pass
-    else:
-        sctypes[typename].append(t)
-
-def _set_array_types():
-    ibytes = [1, 2, 4, 8, 16, 32, 64]
-    fbytes = [2, 4, 8, 10, 12, 16, 32, 64]
-    for bytes in ibytes:
-        bits = 8*bytes
-        _add_array_type('int', bits)
-        _add_array_type('uint', bits)
-    for bytes in fbytes:
-        bits = 8*bytes
-        _add_array_type('float', bits)
-        _add_array_type('complex', 2*bits)
-    _gi = dtype('p')
-    if _gi.type not in sctypes['int']:
-        indx = 0
-        sz = _gi.itemsize
-        _lst = sctypes['int']
-        while (indx < len(_lst) and sz >= _lst[indx](0).itemsize):
-            indx += 1
-        sctypes['int'].insert(indx, _gi.type)
-        sctypes['uint'].insert(indx, dtype('P').type)
-_set_array_types()
-
-
-# Add additional strings to the sctypeDict
-_toadd = ['int', ('float', 'double'), ('complex', 'cdouble'), 
-          'bool', 'object',
-          'str', 'bytes', ('a', 'bytes_'),
-          ('int0', 'intp'), ('uint0', 'uintp')]
-
-for name in _toadd:
-    if isinstance(name, tuple):
-        sctypeDict[name[0]] = allTypes[name[1]]
-    else:
-        sctypeDict[name] = allTypes['%s_' % name]
-
-del _toadd, name
+for k, v in typeinfo.items():
+    concrete_type = v.type
+    allTypes[k] = concrete_type
+    sctypeDict[k] = concrete_type
+
+_aliases = {
+    "double": "float64",
+    "cdouble": "complex128",
+    "single": "float32",
+    "csingle": "complex64",
+    "half": "float16"
+}
+
+for k, v in _aliases.items():
+    sctypeDict[k] = allTypes[v]
+    allTypes[k] = allTypes[v]
+
+# extra aliases are added only to `sctypeDict`
+# to support dtype name access, such as`np.dtype("float")`
+_extra_aliases = {  
+    "bool": "bool_",
+    "float": "float64",
+    "complex": "complex128",
+    "object": "object_",
+    "bytes": "bytes_",
+    "int": "intp",
+    "long": "intp",
```

Comment:
This looks wrong.  `long` has  nothing to do with `intp`, form the C side, you currently get the "current" name `int_`.  You could change that to long on the C-side though (since `int_` is not a nice name).  `int` is probably "default integer" which is `int_` which is `long`.

Once we change the default int, that will need a bit of shuffling around, although if we rename to `long` on the C-side, we can probably do all of that here in Python.

`intp` is its own thing, its included on the C-side dict I provided.